### PR TITLE
Add Resize-18: Antialiasing, axes and keep_aspect_ratio_policy

### PR DESCRIPTION
**Pull request by jantonguirao**
Signed-off-by: Joaquin Anton <janton@nvidia.com>

**Description**
This PR extends Resize with new functionality:
- New boolean parameter `antialias`, which enables an antialiasing filter when downscaling with linear and cubic interpolation modes.

- New `axes` argument, allowing to select a subset of dimensions that `sizes`, `roi` and `scales` refer to.
- New `keep_aspect_ratio_policy` argument, determining how to interpret the value of `sizes`. Three policies supported:
    * "stretch" (default): The output is resized to the target size, disregarding the original aspect ratio
    * "not_larger" : The output is resized so that no extent is larger than the provided size, while keeping its aspect ratio
    * "not_smaller" : The output is resized so that no extent is smaller than the provided size, while keeping its aspect ratio

**Motivation and Context**
- Why is this change required? What problem does it solve?

* Some libraries apply a downscaling antialiasing (e.g. PIL) while others don't (e.g. OpenCV). Having such an option allows the user to be aligned with either. This options will be crucial to be able to port the data preprocessing pipeline from image classification models (resnet), which is currently described as a Python function using PIL resize. For more background see: https://github.com/onnx/working-groups/blob/main/preprocessing/notebooks/resize-antialias/resize-antialias.ipynb
* Previous version of resize forces the user to provide a scale or extent for dimensions that are not meant to be resize. This makes it hard to use in some contexts, such as image resize, where the "channel" dimension should not be resized.
* Resizing data while keeping aspect ratio is currently hard to achieve, requiring a complicated pipeline for a trivial and quite common operation. We can benefit to have such "keep aspect ratio" semantics, for example for image vision applications.

### CODE REVIEW

File: onnx/defs/tensor/defs.cc [hunk 1]
```diff
@@ -2480,14 +2502,34 @@ ONNX_OPERATOR_SET_SCHEMA(
...
         .Attr(
             "extrapolation_value",
             "When coordinate_transformation_mode is \"tf_crop_and_resize\" and x_original is outside the range [0, length_original - 1], this value is used as the corresponding output value. Default is 0.0f.",
             AttributeProto::FLOAT,
             static_cast<float>(0))
+        .Attr(
+            "antialias",
+            "If set to 1, \"linear\" and \"cubic\" interpolation modes will use an antialiasing filter when downscaling. "
+            "Antialiasing is achieved by stretching the resampling filter by a factor max(1, 1 / scale), which means that when downsampling, more input pixels contribute to an output pixel.",
+            AttributeProto::INT,
```
Comment (mzient) [inline onnx/defs/tensor/defs.cc:2517]: No boolean?

File: onnx/defs/tensor/defs.cc [hunk 2]
```diff
@@ -2480,14 +2502,34 @@ ONNX_OPERATOR_SET_SCHEMA(
...
             AttributeProto::STRING,
             std::string("round_prefer_floor"))
         .Attr(
             "extrapolation_value",
             "When coordinate_transformation_mode is \"tf_crop_and_resize\" and x_original is outside the range [0, length_original - 1], this value is used as the corresponding output value. Default is 0.0f.",
             AttributeProto::FLOAT,
             static_cast<float>(0))
+        .Attr(
+            "antialias",
+            "If set to 1, \"linear\" and \"cubic\" interpolation modes will use an antialiasing filter when downscaling. "
```
Comment (mzient) [inline onnx/defs/tensor/defs.cc:2515]: This wording is likely to go stale when a new interpolation method is added. We could say that it is not applicable to "nearest" interpolation method.

File: onnx/defs/tensor/defs.cc (hunk 1 above)
Comment (jantonguirao) [reply onnx/defs/tensor/defs.cc:2517]: No, boolean is not available. INT is typically used for boolean flags, from what I can see in the codebase

File: onnx/defs/tensor/defs.cc (hunk 2 above)
Comment (jantonguirao) [reply onnx/defs/tensor/defs.cc:2515]: That is true. On the other hand, there is a chance that a new interpolation method is added that doesn't support antialiasing.

**Comment by p-wysocki** [comment]
Is `not_smaller` option for `keep_aspect_ratio_policy` necessary? Neither [PyTorch](https://pytorch.org/vision/main/generated/torchvision.transforms.Resize.html) nor [TensorFlow](https://www.tensorflow.org/api_docs/python/tf/image/resize) has an equivalent.

[Adding new operator docs](https://github.com/onnx/sigs/blob/2750e7eb0878d32bb369d9badd8bcf06ebe92a5f/operators/docs/AddNewOpV3.md?plain=1#L72-L73) states that each operator needs to be tested against an existing backend - are there tests for comparing proposed antialiasing functionality?

**Comment by jantonguirao** [comment]
@p-wysocki 
> Is `not_smaller` option for `keep_aspect_ratio_policy` necessary? Neither [PyTorch](https://pytorch.org/vision/main/generated/torchvision.transforms.Resize.html) nor [TensorFlow](https://www.tensorflow.org/api_docs/python/tf/image/resize) has an equivalent.
`not_smaller` offers a similar behavior as [PyTorch Resize](https://pytorch.org/vision/main/generated/torchvision.transforms.Resize.html) when providing a single number as size. In such case Pytorch takes the shorter dimension and scales it to the desired size, scaling the other spatial dimension by the same factor.
The difference is that this proposal allows you to set different minimum or maximum sizes for each dimension individually.

`not_larger` would be equivalent to resizing the largest dimension to the desired size. This is close enough to [TensorFlow resize_with_pad](https://www.tensorflow.org/api_docs/python/tf/image/resize_with_pad). This TF operator resizes to fit the desired canvas size, padding with zeros those dimensions smaller dimensions instead of distorting the aspect ratio. This would be equivalent to Resize with `not_larger` plus a pad step.

> [Adding new operator docs](https://github.com/onnx/sigs/blob/2750e7eb0878d32bb369d9badd8bcf06ebe92a5f/operators/docs/AddNewOpV3.md?plain=1#L72-L73) states that each operator needs to be tested against an existing backend - are there tests for comparing proposed antialiasing functionality?

I have added tests comparing to a handwritten reference python code. I wrote the python reference code and compared against it in tests.

File: onnx/defs/tensor/defs.cc (hunk 1 above)
Comment (gramalingam) [reply onnx/defs/tensor/defs.cc:2517]: That's correct.

File: onnx/defs/tensor/defs.cc [hunk 3]
```diff
@@ -2520,8 +2564,9 @@ ONNX_OPERATOR_SET_SCHEMA(
         .Input(
             3,
             "sizes",
-            "The size of the output tensor. The number of elements of 'sizes' should be the same as the"
-            " rank of input 'X'. Only one of 'scales' and 'sizes' can be specified.",
+            "Target size of the output tensor. Its interpretation depends on the 'keep_aspect_ratio_policy' value."
```
Comment (gramalingam) [inline onnx/defs/tensor/defs.cc:2567]: I assume that the aspect-ratio-policy applies whether scales or sizes is specified?

Comment (jantonguirao) [reply onnx/defs/tensor/defs.cc:2567]: In my opinion, `aspect-ratio-policy` doesn't make sense for scales. If you intend to keep aspect ratio then you set the scale to the same value for all the relevant dimensions. 
Example: 
- Input (300, 200, 3)
You want to double the height, while keeping aspect ratio between height and width.
- Scales (2, 2, 1)

Comment (gramalingam) [reply onnx/defs/tensor/defs.cc:2282]: One minor consideration might be to choose whichever makes implementations (of the op) simpler, but it is probably not a significant difference in this case. So, I am fine with your choice, but it would help avoid any confusion/ambiguity if we say this explicitly.

File: onnx/defs/tensor/defs.cc [hunk 4]
```diff
@@ -2145,40 +2145,62 @@ ONNX_OPERATOR_SET_SCHEMA(
...
-if coordinate_transformation_mode is "half_pixel", <br/>
-x_original = (x_resized + 0.5) / scale - 0.5, <br/>
+if coordinate_transformation_mode is `"tf_crop_and_resize"`, <br/>
+`x_original = length_resized > 1 ? start_x * (length_original - 1) + x_resized * (end_x - start_x) * (length_original - 1) / (length_resized - 1) : 0.5 * (start_x + end_x) * (length_original - 1)`
+.)DOC";
 
-if coordinate_transformation_mode is "pytorch_half_pixel", <br/>
-x_original = length_resized > 1 ? (x_resized + 0.5) / scale - 0.5 : 0, <br/>
+static const char* Resize_attr_keep_aspect_ratio_policy_doc = R"DOC(
+This attribute describes how to interpret the `sizes` input with regard to keeping the original aspect ratio of the input. <br/>
```
Comment (gramalingam) [inline onnx/defs/tensor/defs.cc:2178]: Now I see this does answer my question about `scales` input. Maybe we can add the line
```
and it is not applicable when the `scales` input is used.
```

File: onnx/defs/tensor/utils.cc [hunk 5]
```diff
@@ -19,6 +19,38 @@ void resizeShapeInferenceHelper(
...
+  std::function<float(float, float)> reduce_f;
+  if (policy == KeepAspectRatioPolicy::NOT_LARGER) {
+    reduce_f = [] (float a, float b) { return std::min(a, b); };
+  } else {
+    reduce_f = [] (float a, float b) { return std::max(a, b); };
+  }
+
+  for (size_t i = 0; i < sizes_data.size(); i++) {
+    int d = axes.empty() ? i : axes[i];
+    if (!input_shape.dim(d).has_dim_value())
```
[the remark this answers is not in the document]
Comment (jantonguirao) [reply onnx/defs/tensor/utils.cc]: You are right. I will fix that.

File: onnx/defs/tensor/defs.cc (hunk 4 above)
Comment (jantonguirao) [reply onnx/defs/tensor/defs.cc]: Done

File: docs/Changelog.md [hunk 6]
```diff
@@ -21124,6 +21124,108 @@ This version of the operator has been available since version 17 of the default
...
+
+This version of the operator has been available since version 17 of the default ONNX operator set.
+
+#### Attributes
+
+<dl>
+<dt><tt>antialias</tt> : int (default is 0)</dt>
+<dd>If set to 1, "linear" and "cubic" interpolation modes will use an antialiasing filter when downscaling. Antialiasing is achieved by stretching the resampling filter by a factor max(1, 1 / scale), which means that when downsampling, more input pixels contribute to an output pixel.</dd>
+<dt><tt>axes</tt> : list of ints</dt>
+<dd>If provided, it specifies a subset of axes that 'roi', 'scales' and 'sizes' refer to. If not provided, all axes are assumed [0, 1, ..., r-1], where r = rank(data). Non-specified dimensions are interpreted as non-resizable. Negative value means counting dimensions from the back. Accepted range is [-r, r-1], where r = rank(data). Behavior is undefined if an axis is repeated.</dd>
```
[the remark this answers is not in the document]
Comment (jantonguirao) [reply docs/Changelog.md:21144]: I've added some more error checking

File: onnx/defs/tensor/utils.cc [hunk 7]
```diff
@@ -88,6 +93,31 @@ void resizeShapeInference(InferenceContext& ctx) {
   const auto& input_shape = getInputShape(ctx, 0);
   auto* output_shape = getOutputShape(ctx, 0);
   const TensorProto* scales = 2 < ctx.getNumInputs() ? ctx.getInputData(2) : nullptr;
+  const TensorProto* sizes  = 3 < ctx.getNumInputs() ? ctx.getInputData(3) : nullptr;
+
+  if ((scales != nullptr) + (sizes != nullptr) != 1) {
```
Comment (gramalingam) [inline onnx/defs/tensor/utils.cc:98]: Unfortunately, this check is not the right one (for the warning message below). This condition checks whether a _constant-value_ is available. Unfortunately, there is no good predefined method/function to check whether an input is specified, but the following should work and be okay:
```cpp
   bool hasScalesInput = (2 < ctx.getNumInputs()) && (ctx.getInputType(2) != nullptr);
```

**Comment by yuanyao-nv** [comment]
Should we be more precise in the ONNX spec about what antialias filter to use for various interpolation methods? For example, by giving the precise mathematical formula in cases of possible ambiguity.

**Comment by jantonguirao** [comment]
> Should we be more precise in the ONNX spec about what antialias filter to use for various interpolation methods? For example, by giving the precise mathematical formula in cases of possible ambiguity.


@yuanyao-nv There is this definition:
```
Antialiasing is achieved by stretching the resampling filter by a factor max(1, 1 / scale), which means that when downsampling, more input pixels contribute to an output pixel.
```
which is basically extending the neighborhood region by the inverse of the scale factor, resampling the interpolation filter weights accordingly. Do you think we should phrase it differently? 

Regarding being mathematically explicit, I feel that for this we would need to express mathematically all the interpolation methods with all the variants, which might be hard to do concisely

**Comment by jantonguirao** [comment]
> Thanks for the reply @jantonguirao . Do you know if this is the same antialias definition used in [TensorFlow](https://www.tensorflow.org/api_docs/python/tf/image/resize) and [PyTorch](https://pytorch.org/vision/main/generated/torchvision.transforms.Resize.html) ? In the TF definition, they mention some hat/tent filter for bilinear.

It seems to be the same filter as those two frameworks.

Regarding Tensorflow:
```becomes a hat/tent filter function with radius 1 when downsampling.```
Yes, this is the same filter. Also known as triangular filter.

Regarding Pytorch:
They mention being close to what PIL does. PIL applies antialiasing with the same approach as I am suggesting here. This is a triangular filter for the case of bilinear interpolation.

I made a demo comparing this proposal against OpenCV (no antialiasing) and Pillow (antialiasing). As you can see, the formulas I am proposing match what Pillow is doing.

**Review by gramalingam — APPROVED** [review]

File: onnx/defs/tensor/utils.cc [hunk 8]
```diff
@@ -54,7 +91,36 @@ void resizeShapeInference(InferenceContext& ctx) {
   }
   const auto& input_shape = getInputShape(ctx, 0);
   auto* output_shape = getOutputShape(ctx, 0);
+
+  bool hasScalesInput = (2 < ctx.getNumInputs()) && (ctx.getInputType(2) != nullptr);
+  bool hasSizesInput = (3 < ctx.getNumInputs()) && (ctx.getInputType(3) != nullptr);
+
+  if (hasScalesInput + hasSizesInput != 1) {
+    fail_shape_inference("Either `sizes` or `scales` must be provided, but not both of them");
```
Comment (jcwchen) [inline onnx/defs/tensor/utils.cc:99]: FYI this line causes shape inference failures with some of ONNX Model Zoo models. Please check this issue: https://github.com/onnx/onnx/issues/4380. Thanks!
